# Part 2: NumPy Exercises

NumPy's object is the homogeneous multidimensional array. In this exercise, you'll learn common operations in NumPy. <br>

Time to complete: 15 mins

## Import NumPy Library

In [1]:
import numpy as np

## 1. Basics
NumPy list can be constructed with np.array(list). 

In [2]:
list1 = [1,2,3,4,5]
x = np.array(list1)

Unlike Python lists, NumPy does not allow the elements of different types. If types do no match, NumPy will upcast if possible.

In [3]:
x = np.array([1,2,3,"4","5"])
x

array(['1', '2', '3', '4', '5'], dtype='<U21')

### Suppose you have regular sales_list and expenses_list as below.

In [4]:
sales_list = [100, 120, 130, 150]
expenses_list = [50, 55, 45, 40]

### Exercise 1.1. Create two new ndarrays. 
`np_sales_list` for sales_list <br>
`np_expenses_list` for expenses_list

In [5]:
# Your answer for Exercise 1.1. (Hint: np.array(list))
np_sales_list = 
np_expenses_list = 

### NumPy Array Attributes
1. `ndim` - Number of dimensions <br>
2. `shape` - Size of each dimension <br>
3. `size` - Number of arrray elements <br>
4. `dtype` - Type of elements in array <br>
5. `itemsize` - size in bytes of each element of the array <br>

### Exercise 1.2. Print all attributes of array `np_sales_list`.

In [6]:
# Your answer for Exercise 1.2.
print(np_sales_list.ndim)

# Print the ndarray attributes (shape, size, dtype, itemsize).

1


### Exericise 1.3. Get the sum of `np_sales_list` and `np_expenses_list`.

In [7]:
# Your answer for Exercise 1.3. For np_sales_list (Hint: use np_list.sum())

500

In [8]:
# Your answer for Exercise 1.3. For np_expenses_list (Hint: use np_list.sum())

190

NumPy allows you to perform arithematic operation over an entire array. Let's calculate our profit from sales and expense for each year.

### Exercise 1.4. Create `np_profits_list` by subtracting `np_expenses_list` from `np_sales_list`.

In [9]:
# Your answer for Exercise 1.4.
np_profits_list = 

In [10]:
np_profits_list

array([ 50,  65,  85, 110])

### Exercise 1.5. Caculate 10% increase of the existing profit for each year. 

In [11]:
# Your answer for Exercise 1.5. (Multiply np_profilts_list with 1.1)

array([ 55. ,  71.5,  93.5, 121. ])

### Exercise 1.6. Get sales data from np_sales_list which is greater than 110.

In [12]:
# Your answer for Exercise 1.6. np_list[condition]

array([120, 130, 150])

### 2. NumPy 2D arrays
Previously you have 2 1-Dimensional arrays. Now let's group them together into 1 single 2-Dimensional arrays.

### Exercise 2.1. Create a new `np_sales_expenses` ndarray with the 2 following ndarrays. With row 0 as the `sales_list`.

In [13]:
sales_list = [100, 120, 130, 150]
expenses_list = [50, 55, 45, 40]

In [14]:
# Your answer for Exercise 2.1. np_list = np.array([list1, list2])
np_sales_expenses = 

In [15]:
np_sales_expenses

array([[100, 120, 130, 150],
       [ 50,  55,  45,  40]])

## Subsetting rows and columns

In this new 2D array, row 0 represents sales and row 1 represents expenses while each column represents each year. Use the following syntax to get value from 2D array.

In [16]:
# np_sales_expenses[row_index, column_index]

### Exericise 2.2. Select sales for year 2 from `np_sales_expenses`, the result value is `120`.

In [17]:
# Your answer for Exercise 2.2.

120

### Exercise 2.3. Slice sales and expenses for year 3 from `np_sales_expenses`, the result value is `[130, 45]`.

In [18]:
# Your answer for Exercise 2.3. (Hint: Slicing syntax np_list[start_index:stop_index])

array([130,  45])

### Exercise 2.4. Get the average expenses for all 4 years from `np_sales_expenses`.

In [19]:
# Your answer for Exercise 2.4. (Hint: expenses is at row 1. Use np_list.mean() to get average.)

47.5